lib/bundler/settings.rb

```diff
@@ -279,24 +289,25 @@ def array_to_s(array)
       array.join(":").tr(" ", ":")
     end
 
-    def set_key(key, value, hash, file)
-      value = array_to_s(value) if is_array(key)
+    def set_key(raw_key, value, hash, file)
+      raw_key = raw_key.to_s
+      value = array_to_s(value) if is_array(raw_key)
 
-      key = key_for(key)
+      key = key_for(raw_key)
 
-      unless hash[key] == value
-        hash[key] = value
-        hash.delete(key) if value.nil?
-        if file
-          SharedHelpers.filesystem_access(file) do |p|
-            FileUtils.mkdir_p(p.dirname)
-            require "bundler/yaml_serializer"
-            p.open("w") {|f| f.write(YAMLSerializer.dump(hash)) }
-          end
-        end
-      end
+      return if hash[key] == value
+
+      hash[key] = value
+      hash.delete(key) if value.nil?
+
+      Validator.validate(raw_key, converted_value(value, raw_key), hash)
```

Comment:
I think it's easier to do it afterward -- it means rules can validate multiple things at once just by looking at the passed-in hash